code/scripting/api/objs/model.cpp

```diff
@@ -243,15 +278,186 @@ ADE_VIRTVAR(Radius, l_Model, "number", "Model radius (Used for collision & culli
+{
+	submodel_h *smh = nullptr;
+
+	if (!ade_get_args(L, "o", l_Submodel.GetPtr(&smh)))
+		return ade_set_error(L, "s", "");
+
+	if (!smh->IsValid())
+		return ade_set_error(L, "s", "");
+
+	if (ADE_SETTING_VAR)
+		LuaError(L, "Setting the submodel offset is not implemented");
+
+	return ade_set_args(L, "s", smh->GetSubmodel()->name);
+}
+
+ADE_VIRTVAR(Offset, l_Submodel, nullptr, "Gets the submodel's offset from its parent submodel", "vector", "The offset vector or a null vector if invalid")
+{
+	submodel_h *smh = nullptr;
+
+	if (!ade_get_args(L, "o", l_Submodel.GetPtr(&smh)))
+		return ade_set_error(L, "o", l_Vector.Set(vmd_zero_vector));
+
+	if (!smh->IsValid())
+		return ade_set_error(L, "o", l_Vector.Set(vmd_zero_vector));
+
+	if (ADE_SETTING_VAR)
+		LuaError(L, "Setting the submodel offset is not implemented");
+
+	return ade_set_args(L, "o", l_Vector.Set(smh->GetSubmodel()->offset));
+}
+
+ADE_VIRTVAR(Radius, l_Submodel, nullptr, "Gets the submodel's radius", "number", "The radius of the submodel or -1 if invalid")
+{
+	submodel_h *smh = nullptr;
+
+	if (!ade_get_args(L, "o", l_Submodel.GetPtr(&smh)))
+		return ade_set_error(L, "f", -1.0f);
+
+	if (!smh->IsValid())
+		return ade_set_error(L, "f", -1.0f);
+
+	if (ADE_SETTING_VAR)
+		LuaError(L, "Setting the submodel radius is not implemented");
+
+	return ade_set_args(L, "f", smh->GetSubmodel()->rad);
+}
+
+ADE_FUNC(getFirstChild, l_Submodel, nullptr, "Gets the first child submodel of this submodel", "submodel", "A submodel, or nil if there is no child")
+{
+	submodel_h *smh = nullptr;
+
+	if (!ade_get_args(L, "o", l_Submodel.GetPtr(&smh)))
+		return ade_set_error(L, "o", l_Submodel.Set(submodel_h()));
+
+	if (!smh->IsValid())
+		return ade_set_error(L, "o", l_Submodel.Set(submodel_h()));
+
+	auto sm = smh->GetSubmodel();
+	if (sm->first_child < 0)
+		return ADE_RETURN_NIL;
+
+	return ade_set_args(L, "o", l_Submodel.Set(submodel_h(smh->GetModel(), sm->first_child)));
+}
+
+ADE_FUNC(getNextSibling, l_Submodel, nullptr, "Gets the next sibling submodel of this submodel", "submodel", "A submodel, or nil if there are no remaining siblings")
+{
+	submodel_h *smh = nullptr;
+
+	if (!ade_get_args(L, "o", l_Submodel.GetPtr(&smh)))
+		return ade_set_error(L, "o", l_Submodel.Set(submodel_h()));
+
+	if (!smh->IsValid())
+		return ade_set_error(L, "o", l_Submodel.Set(submodel_h()));
+
+	auto sm = smh->GetSubmodel();
+	if (sm->next_sibling < 0)
+		return ADE_RETURN_NIL;
+
+	return ade_set_args(L, "o", l_Submodel.Set(submodel_h(smh->GetModel(), sm->next_sibling)));
+}
+
+ADE_FUNC(isValid, l_Submodel, nullptr, "True if valid, false or nil if not", "boolean", "Detects whether handle is valid")
+{
+	submodel_h *smh;
+	if (!ade_get_args(L, "o", l_Submodel.GetPtr(&smh)))
+		return ADE_RETURN_NIL;
+
+	return ade_set_args(L, "b", smh->IsValid());
+}
+
+
+//**********HANDLE: modelsubmodels
+ADE_OBJ(l_ModelSubmodels, modelsubmodels_h, "modelsubmodels_h", "Array of submodels");
```

Comment:
This should not be its own handle type.
Arrays of a custom type are otherwise handled using sublibraries in other places, such as the mission library (see [Mission Library](https://github.com/scp-fs2open/fs2open.github.com/blob/master/code/scripting/api/libs/mission.cpp#L160...L185), just that instead of throwing away the first object, capture it as l_Model, as is done in most other cases where we use indexers)

Reply:
I patterned this after the other arrays in model.cpp, but I'll take a look at the mission library

Reply:
IIRC the only place that does it like that are the model textures, and I _really_ don't like these as well.
We should avoid creating more of these, as it introduces additional types that have no use apart from wrapping functions.